pkg/proxy/iptables/proxier_test.go

```diff
@@ -709,6 +711,87 @@ func errorf(msg string, rules []iptablestest.Rule, t *testing.T) {
 	t.Errorf("%v", msg)
 }
 
+// findAllMatches takes an array of lines and a pattern with one parenthesized group, and
+// returns a sorted array of all of the unique matches of the parenthesized group.
+func findAllMatches(lines []string, pattern string) []string {
+	regex := regexp.MustCompile(pattern)
+	allMatches := sets.NewString()
+	for _, line := range lines {
+		match := regex.FindStringSubmatch(line)
+		if len(match) == 2 {
+			allMatches.Insert(match[1])
+		}
+	}
+	return allMatches.List()
+}
+
+// moveMatchingLines moves lines that match pattern from input to output
+func moveMatchingLines(pattern string, input, output []string) ([]string, []string) {
+	var newIn []string
+	regex := regexp.MustCompile(pattern)
+	for _, line := range input {
+		if regex.FindString(line) != "" {
+			output = append(output, line)
+		} else {
+			newIn = append(newIn, line)
+		}
+	}
+	return newIn, output
+}
+
+// sortIPTablesRules sorts `iptables-restore` output so as to not depend on the order that
+// Services get processed in, while preserving the relative ordering of related rules.
+func sortIPTablesRules(t *testing.T, ruleData string) string {
+	tables := strings.Split(ruleData, "COMMIT\n")
+	if len(tables) != 3 || tables[2] != "" {
+		t.Fatalf("wrong number of tables (%d) in ruleData\n%s", len(tables), ruleData)
+	}
+	tables = tables[:2]
+	var output []string
+
+	for _, table := range tables {
+		lines := strings.Split(strings.Trim(table, "\n"), "\n")
+
+		// Move "*TABLENAME" line
+		lines, output = moveMatchingLines(`^\*`, lines, output)
+
+		// Move and sort ":CHAINNAME" lines
+		for _, chainName := range findAllMatches(lines, `^(:[^ ]*) `) {
+			lines, output = moveMatchingLines(chainName, lines, output)
+		}
+
+		// Move KUBE-NODEPORTS rules for each service, sorted by service name
```

Comment:
What makes this sorted?

Reply:
`findAllMatches` returns the matches in sorted order, so we start with a sorted list of services and then move the NODEPORTS rules for them one by one